src/datamodel/builder/UMessage.cpp

```diff
@@ -10,3 +10,230 @@
+        (attributes_.type() == v1::UMessageType::UMESSAGE_TYPE_RESPONSE)) {
+        if (priority < v1::UPriority::UPRIORITY_CS4) {
+            throw std::out_of_range("Priority value is out of range, for req/resp should be > CS4");
+        }
+    }
+
+    attributes_.set_priority(priority);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withTtl(std::chrono::milliseconds ttl) {
+    if ((ttl.count() <= 0) || (ttl.count() > std::numeric_limits<uint32_t>::max())) {
+        throw std::out_of_range("TTL value is out of range");
+    }
+
+    attributes_.set_ttl(static_cast<uint32_t>(ttl.count()));
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withToken(const std::string& token) {
+    if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+        throw std::domain_error("Token can only be set on a request message");
+    }
+
+    attributes_.set_token(token);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPermissionLevel(uint32_t level) {
+    if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+        throw std::domain_error("Permission level can only be set on a request message");
+    }
+
+    attributes_.set_permission_level(level);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withCommStatus(v1::UCode code) {
+    if (attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_RESPONSE) {
+        throw std::domain_error("CommStatus can only be set on a response message");
+    }
+
+    if ( ! (v1::UCode_IsValid(code)) ) {
+        throw std::out_of_range("UCode is out of range" + std::to_string(code));
+    }
+
+    if (code != v1::UCode::OK) {
+        attributes_.set_commstatus(code);
+    }
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPayloadFormat(v1::UPayloadFormat format) {
+    if ((format < v1::UPayloadFormat_MIN) ||
+        (format > v1::UPayloadFormat_MAX)) {
+        throw std::out_of_range("Payload format value is out of range");
+    }
+
+    attributes_.set_payload_format(format);
+
+    return *this;
+}
+
+v1::UMessage UMessageBuilder::build() const {
+    v1::UMessage message;
+    if ((attributes_.payload_format() >= v1::UPayloadFormat_MIN) &&
+        (attributes_.payload_format() <= v1::UPayloadFormat_MAX)) {
+
+        throw UnexpectedFormat("payload format did not match the one set with withPayloadFormat()");
+    }
+
+    message.set_allocated_attributes(new v1::UAttributes(attributes_));
+
+    return message;
+}
+
+v1::UMessage UMessageBuilder::build(builder::Payload&& payload) const {
+    v1::UMessage message;
+
+    message.set_allocated_attributes(new v1::UAttributes(attributes_));
+
+    return message;
+}
+
+UMessageBuilder::UMessageBuilder(v1::UMessageType msgType,
+                v1::UUri&& source,
+                v1::UUri&& sink,
+                v1::UUID&& request_id) {
+
+    attributes_.set_type(msgType);
+
+    v1::UUri* sourcePtr = new v1::UUri(std::move(source));
```

Comment:
Please avoid using set_allocated method because it needs to allocate memory using new operator. Try using mutable instead. Here is the code:

UMessageBuilder::UMessageBuilder(v1::UMessageType msgType,
                v1::UUri&& source,
                v1::UUri&& sink,
                v1::UUID&& request_id) {

    attributes_.set_type(msgType);

    *attributes_.mutable_source() = std::move(source);

    if (sink.IsInitialized()) {
        *attributes_.mutable_sink() = std::move(sink);
    }

    attributes_.set_priority(v1::UPriority::UPRIORITY_CS1);

    if (request_id.IsInitialized()) {
        *attributes_.mutable_reqid() = std::move(request_id);
    }

    attributes_.set_payload_format(static_cast<v1::UPayloadFormat>(v1::UPayloadFormat_MIN - 1));
}

Reply:
@gregmedd @billpittman, what do you think about this ?

Reply:
No super strong opinion really.  Are the objects being copied into the function and then again when we set their values with set_allocated_**** ?  So two copies, then if so, I think it would be wise to use the mutable version.

Reply:
I agree with @pranavishere2. Manual memory management (e.g. via `new`) should _never_ be done without _very good reason_.

For the record, we are unlikely to ever have a good reason in up-cpp.

Reply:
Done